coroutines-codelab/finished_code/src/test/java/com/example/android/kotlincoroutines/main/utils/MainCoroutineRule.kt

```diff
@@ -0,0 +1,24 @@
+package com.example.android.kotlincoroutines.main.utils
+
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.setMain
+import org.junit.rules.TestWatcher
+import org.junit.runner.Description
+
+class MainCoroutineRule(val dispatcher: TestCoroutineDispatcher = TestCoroutineDispatcher()):
+        TestWatcher(),
+        TestCoroutineScope by TestCoroutineScope(dispatcher) {
```

Comment:
Does this need to implement TestWatcher and TestCoroutineScope? I wouldn't do that and instead create extension functions if it's needed as this code can be copy-pasted. 

This favours inheritance vs composition 👎 

The MainRule *has* a dispatcher, the MainRule *shouldn't be* a Dispatcher. Same as when we used to recommended ViewModel to implement CoroutineScope 👎

Reply:
In this case I'm in favor of the inheritance because the point of this rule is to give access to a JUnit4-aware CoroutineScope to the caller.

It's maybe more properly named "Junit4CoroutineScope"

Creating extension functions involves both providing runBlockingTest as well as the entire DelayController interface. AFAIK there's not a way to say "implement delegate extensions."

Note, when I was reviewing Lyla's slides yesterday I saw that the runBlockingTest in plaid is already out of sync with the published API, and it won't keep in sync with the new timeout interface when it lands.

Reply:
In this case I'm in favor of the inheritance because the point of this rule is to give access to a JUnit4-aware CoroutineScope to the caller. This is different than the ViewModel case where the intention is to provide a CoroutineScope inside the ViewModel (but not expose it to callers).

Creating equivalent extension functions involves both providing `runBlockingTest` as well as delegating the `DelayController` interface – which sounds hard to maintain.

Reply:
As a note: if it were possible to not extend TestWatcher without making invocation hard that would be preferred.

Reply:
Note, copying this code as is for use in the third lesson of the testing course.

Reply:
Add @ExperimentalCoroutinesApi to avoid lint warnings

Reply:
Definitely, removing TestWatcher is impossible haha (sorry, my bad).

Should we consider changing the name to represent that it's a scope? E.g. `MainCoroutineScopeRule`, I'm conscious of people looking at tests (without knowing the rule existence) and not being able to figure out why a rule can do all that.

Reply:
I've been wondering the same thing for a while – do you like that name? If so renames incoming :)>

Reply:
Sure! I wouldn't mind that one :)

Reply:
+1 on rename then. I'm going to keep the `Main` prefix since it only does main, but add a comment saying that it should inject other injectable dispatchers if using something like Dagger 2.